server/src/main/java/org/apache/iotdb/db/wal/buffer/WALBuffer.java

```diff
@@ -0,0 +1,426 @@
+      }
+    }
+  }
+
+  /**
+   * This view uses workingBuffer lock-freely because workingBuffer is only updated by
+   * serializeThread and this class is only used by serializeThread.
+   */
+  private class ByteBufferView implements IWALByteBufferView {
+    private void ensureEnoughSpace(int bytesNum) {
+      if (workingBuffer.remaining() < bytesNum) {
+        rollBuffer();
+      }
+    }
+
+    private void rollBuffer() {
+      syncWorkingBuffer();
+    }
+
+    @Override
+    public void put(byte b) {
+      ensureEnoughSpace(Byte.BYTES);
+      workingBuffer.put(b);
+    }
+
+    @Override
+    public void put(byte[] src) {
+      int offset = 0;
+      while (true) {
+        int leftCapacity = workingBuffer.remaining();
+        int needCapacity = src.length - offset;
+        if (leftCapacity >= needCapacity) {
+          workingBuffer.put(src, offset, needCapacity);
+          break;
+        } else {
+          workingBuffer.put(src, offset, leftCapacity);
+          offset += leftCapacity;
+          rollBuffer();
+        }
+      }
+    }
+
+    @Override
+    public void putChar(char value) {
+      ensureEnoughSpace(Character.BYTES);
+      workingBuffer.putChar(value);
+    }
+
+    @Override
+    public void putShort(short value) {
+      ensureEnoughSpace(Short.BYTES);
+      workingBuffer.putShort(value);
+    }
+
+    @Override
+    public void putInt(int value) {
+      ensureEnoughSpace(Integer.BYTES);
+      workingBuffer.putInt(value);
+    }
+
+    @Override
+    public void putLong(long value) {
+      ensureEnoughSpace(Long.BYTES);
+      workingBuffer.putLong(value);
+    }
+
+    @Override
+    public void putFloat(float value) {
+      ensureEnoughSpace(Float.BYTES);
+      workingBuffer.putFloat(value);
+    }
+
+    @Override
+    public void putDouble(double value) {
+      ensureEnoughSpace(Double.BYTES);
+      workingBuffer.putDouble(value);
+    }
+  }
+
+  /** Notice: this method only called when buffer is exhausted by SerializeTask. */
+  private void syncWorkingBuffer() {
+    switchIdleBufferToWorking();
+    syncBufferThread.submit(new SyncBufferTask(false));
+  }
+
+  /** Notice: this method only called at the last of SerializeTask. */
+  private void fsyncWorkingBuffer(List<WALFlushListener> fsyncListeners) {
+    switchIdleBufferToWorking();
+    syncBufferThread.submit(new SyncBufferTask(true, fsyncListeners));
+  }
+
+  // only called by serializeThread
+  private void switchIdleBufferToWorking() {
+    buffersLock.lock();
+    try {
+      while (idleBuffer == null) {
+        idleBufferReadyCondition.await();
+      }
+      syncingBuffer = workingBuffer;
+      workingBuffer = idleBuffer;
+      workingBuffer.clear();
+      idleBuffer = null;
+    } catch (InterruptedException e) {
+      logger.warn("Interrupted When waiting for available working buffer.");
+      Thread.currentThread().interrupt();
+    } finally {
+      buffersLock.unlock();
+    }
+  }
+  // endregion
```

Comment:
delete

Reply:
Get lock at the [beginning of this method](https://github.com/apache/iotdb/pull/5320/files#diff-867c4e44ff59fdd079abe7337e64b9774fd55bffb89ec743bb42db0e2f7fab77R283-R285), so it's unsafe to delete lock release code.